#' @title Run DESeq2 for computed sample combinations
#'
#' @description
#' \code{erssa_deseq2} function runs DESeq2 Wald test to identify
#' differentially expressed (DE) genes for each sample combination computed by
#' \code{comb_gen} function. A gene or transcript is considered to be
#' differentially expressed by defined padj (Default=0.05) and log2FoldChange
#' (Default=1) values. As an option, the function can also save the DESeq2 result
#' tables as csv files in the current working directory.
#'
#' @details
#' The main function calls DESeq2 functions to perform Wald test for each
#' computed combinations generated by \code{comb_gen}. In all tests, the
#' pair-wise test sets the condition defined in the object "control" as the
#' control condition.
#'
#' In typical usage, after each test, the list of differentially expressed genes
#' are filtered by padj and log2FoldChange values and only the filtered gene or
#' transcript names are saved for further analysis. However, it is also possible
#' to save all of the generated result tables to the drive for analysis outside
#' the scope of this package.
#'
#' @param count_table.filtered Count table pre-filtered to remove non- to low- expressing genes or transcripts. Can be the output of \code{count_filter} function.
#' @param combinations List of combinations that is produced by \code{comb_gen} function.
#' @param condition_table A condition table with two columns and each sample as a row. Column 1 contains sample names and Column 2 contains sample condition (e.g. Control, Treatment).
#' @param control One of the condition names that will serve as control.
#' @param cutoff_padj The cutoff in padj for DE consideration. Genes with lower padj pass the cutoff. Default = 0.05.
#' @param cutoff_Abs_logFC the cutoff in abs(log2FoldChange) for differential expression consideration. Genes with higher abs(log2FoldChange) pass the cutoff. Default = 1.
#' @param save_table Boolean. When set to TRUE, function will, in addition, save the generated DESeq2 result table as csv files. The files are saved on the drive in the working directory in a new folder named "ERSSA_DESeq2_table". Tables are saved separately by the replicate level. Default = FALSE.
#'
#' @return A list of list of vector. Top list contains elements corresponding to replicate levels. Each child list contains elements corresponding to each combination at the respective replicate level. The child vectors contain differentially expressed gene or transcript names.
#'
#' @examples
#' deg = erssa_deseq2(count_table.filtered, combinations, condition_table, control='heart')
#'
#' @export

erssa_deseq2 = function(count_table.filtered, combinations, condition_table, control, cutoff_padj = 0.05, cutoff_Abs_logFC = 1, save_table=FALSE){

  DE_genes = list() #list to save DEGs at all replicate levels
  cat(paste0('Start DESeq2 Wald test with padj cutoff = ',cutoff_padj, ', Abs(log2FoldChange) cutoff = ', cutoff_Abs_logFC,'\n'))
  cat(paste0('Save results tables to drive: ', save_table,'\n'))

  if (save_table==TRUE){
    #create dir to save results
    folder_path = './ERSSA_DESeq2_table'
    dir.create(folder_path, showWarnings = FALSE)
    setwd(folder_path)
  }

  #loop through each replicate level
  for (rep_level in names(combinations)){

    DE_genes_rl = list() #list to save DEGs at one replicate level
    comb_rl = combinations[rep_level][[1]]

    if (save_table==TRUE){
      #create dir to save results
      folder_path = paste0('./',rep_level)
      dir.create(folder_path, showWarnings = FALSE)
      setwd(folder_path)
    }

    #loop through each combination at given replicate level
    for (index in seq(1, length(comb_rl))){

      #combination string to vector
      comb_samples_i = strsplit(comb_rl[index], ';')[[1]]

      #parse out sample counts
      count_table.filtered_i = count_table.filtered[,colnames(count_table.filtered) %in% comb_samples_i]

      #DESeq2
      group = sapply(colnames(count_table.filtered_i), function(x)
        condition_table$condition[which(condition_table$name==x)])
      group=as.data.frame(group)
      dds = DESeq2::DESeqDataSetFromMatrix(countData = count_table.filtered_i,
                                         colData = group,
                                         design = ~ group)
      dds$group = stats::relevel(dds$group, ref = control)
      dds = DESeq2::DESeq(dds, quiet=TRUE)
      resLFC = DESeq2::lfcShrink(dds, coef=2)

      res_cutoff = resLFC[which(resLFC$padj < cutoff_padj & abs(resLFC$log2FoldChange) > cutoff_Abs_logFC),]

      #save list of DE genes to list
      DE_genes_rl[[paste0('comb_',index)]] = rownames(res_cutoff)

      #save entire result table to drive
      if (save_table==TRUE){
        utils::write.csv(resLFC,paste0('ERSSA_DESeq2_',rep_level,'_comb',index,'.csv'))
      }

      cat(paste0(rep_level,'; combination_',index,' | done\n'))
    }

    #add DE gene lists to main list
    DE_genes[[rep_level]]=DE_genes_rl
    if (save_table==TRUE){
      setwd('../')
    }

  }

  return(DE_genes)
}









