#' @title Empirical RNA-seq Sample Size Analysis
#'
#' @description
#' ERSSA is a package designed to test whether an currently available RNA-seq dataset has sufficient biological replicates to detect a majority of differentially expressed (DE) genes between two conditions. Base on the number of biological replicates available, the algorithm subsamples at step-wise replicate levels and uses existing differentially expression analysis softwares (e.g. edgeR and DESeq2) to identify the number of DE genes. This process is repeated for a given number of times with unique combinations of samples to generate a distribution of DE genes at each replicate level. Compare to existing RNA-seq sample size analysis algorithms, ERSSA does not rely on any a priori assumptions about the dataset, but rather uses an user-supplied pilot RNA-seq dataset to determine whether the current replicate level is sufficient to detect a majority of DE genes.
#'
#' @details
#' \code{erssa} function is a wrapper that calls several ERSSA functions in sequence. For additional description of the functions called, please see their respective manual.
#'
#' For the majority of current RNA-seq analysis, RNA-seq samples are aligned to the reference, followed by running quantification packages to generate count tables. ERSSA can then take the unfiltered count table and remove non- to low-expressing genes with count_filter function. Alternatively, a filtered count table can be supplied and filtering will be skipped.
#'
#' Next, unique combinations of samples at various biological replicate levels will be generated by comb_gen function and passed to a differential expression analysis software for statistical testing for DE genes. The pipeline currently supports edgeR and DESeq2, but additional software support can be easily added.
#'
#' Finally, the result trend in DE gene identification can be visualized by plot (ggplot2_dotplot) and the user can determine whether the desirable level of DE gene discovery has been reached.
#'
#' At the default setting, the results of statistical tests are not saved, only the list of DE genes is. However, all of the test results can be optionally saved for further analysis.
#'
#' @param count_table A RNA-seq count matrix with genes on each row and samples on each column. If count_table has already been filtered to remove non- or low-expressing genes, then counts_filtered argument should be changed to TRUE.
#' @param condition_table A condition table with two columns and each sample as a row. Column 1 contains sample names and Column 2 contains sample conditions (e.g. Control, Treatment).
#' @param DE_software The name of DE analysis software to use. Current options include "edgeR" and "DESeq2". Default to "edgeR".
#' @param DE_ctrl_cond The name of control condition in the comparison. Must be one of the two conditions in the condition table.
#' @param DE_cutoff_stat The cutoff in FDR or adjusted p-value used to determine whether a gene is differentially expressed. Genes with lower FDR or adjusted p-value pass the cutoff. Default = 0.05.
#' @param DE_cutoff_Abs_logFC The cutoff in abs(log2FoldChange) for differential expression consideration. Genes with higher abs(log2FoldChange) pass the cutoff. Default = 1.
#' @param DE_save_table Boolean. The results of differential expression tests can be saved to the drive for further analysis. Default setting does not save the results to save drive space. Default = FALSE.
#' @param path The path to which the plots and results will be saved. Default to current working directory.
#' @param filter_cutoff The average CPM threshold set for filtering genes. Default to 1.
#' @param counts_filtered Boolean. Whether count table has already been filtered. Default = FALSE with the function run filtering by average CPM at the cutoff specified by filter_cutoff value.
#' @param comb_gen_seed To generate unique combinations of samples at each replicate level, a random process is employed. The seed can be set for reproducible combination generation. Default to no seed.
#' @param comb_gen_repeat The number of maximum unique combinations to generate at each replicate level. More tests will be performed with a bigger value, but run time also increases linearly. Default set to 20 unique combinations at maximum.
#'
#' @return A list of objects generated during the analysis is returned:
#' \itemize{
#'  \item{count_table.filtered}{filtered count table}
#'  \item{samp.name.comb}{the samples involved in each statistical test}
#'  \item{list.of.DE.genes}{list of DE genes in each statistical test}
#'  \item{gg.dotplot.obj}{a list of objects that can be used to recreate the dot plot. Can be used for further modification of the plot}
#' }
#'
#' @author Zixuan Shao, \email{Zixuanshao.zach@@gmail.com}
#'
#' @examples
#' #load data
#' data(condition_table, package = "ERSSA")
#' data(count_table, package = "ERSSA")
#'
#' #run erssa main wrapper script
#' ssa = erssa(count_table, condition_table, DE_ctrl_cond='heart')
#'
#' @export


erssa = function(count_table=NULL, condition_table=NULL, DE_software='edgeR', DE_ctrl_cond=NULL, DE_cutoff_stat = 0.05, DE_cutoff_Abs_logFC = 1, DE_save_table=FALSE, path='.', filter_cutoff=1, counts_filtered=FALSE, comb_gen_seed=NULL, comb_gen_repeat=20){

  #check all required arguments supplied
  if (is.null(count_table)){
    stop('Missing required count_table argument')
  } else if (is.null(condition_table)){
    stop("Missing required condition_table argument")
  } else if (is.null(DE_ctrl_cond)){
    stop("Missing name of control condition in comparison")
  }


  #filter the count table by average CPM value cutoff
  #default is to filter the count_table with cutoff=1
  if (counts_filtered==FALSE){
    count_table.filtered = ERSSA::count_filter(count_table=count_table, cutoff=filter_cutoff, path=path)
  } else if (counts_filtered==TRUE){
    count_table.filtered = count_table
  } else {
    stop("counts_filtered is a boolean variable!")
  }


  #generate sample name combinations at various replicate levels
  combinations = ERSSA::comb_gen(condition_table=condition_table, n_repetition=comb_gen_repeat, seed=comb_gen_seed, path=path)


  #run DE software to generate DE gene list
  if (DE_software=='edgeR'){
    deg = ERSSA::erssa_edger(count_table.filtered=count_table.filtered, combinations=combinations, condition_table=condition_table, control=DE_ctrl_cond, cutoff_stat = DE_cutoff_stat, cutoff_Abs_logFC = DE_cutoff_Abs_logFC, save_table=DE_save_table, path=path)

  } else if (DE_software=='DESeq2'){
    deg = ERSSA::erssa_deseq2(count_table.filtered=count_table.filtered, combinations=combinations, condition_table=condition_table, control=DE_ctrl_cond, cutoff_stat = DE_cutoff_stat, cutoff_Abs_logFC = DE_cutoff_Abs_logFC, save_table=DE_save_table, path=path)
  }

  #plot number of DE genes as dot plot and boxplot
  gg_dot = ERSSA::ggplot2_dotplot(deg=deg, path=path)


  return(list(count_table.filtered=count_table.filtered, samp.name.comb=combinations, list.of.DE.genes=deg, gg.dotplot.obj = gg_dot))
}



