#' @title Plot number of DE genes that is common across combinations
#'
#' @description
#' \code{ggplot2_intersectPlot} function generates and plots the list of
#' differentially expressed (DE) genes that are commonly found in all
#' combinations at any particular replicate level. Often in small-scale
#' RNA-seq experiments, the inclusion or exclusion of any paricular sample can
#' result in a highly different list of DE genes. To reduce the influence of any
#' particular sample in the entire dataset analysis, it may be desirable to
#' identify the list of DE genes that are enriched regardless of any specific
#' sample(s) inclusion. This approach may be most useful analyzing the list of
#' common DE genes at the greatest possible replicate to take advantage of the
#' robust feature as well as employing typically the longest list of DE genes.
#'
#' @details
#' Similar to how increasing number of detected DE genes can be found with more
#' biological replicates, the list of common DE genes is expected to increase
#' with more replicates. This eventually levels off as majority of DE genes have
#' been found.
#'
#' Another similar analytical approach is to repeat DE analysis with all
#' combinations of samples with one sample left out. In typical datasets with
#' equal number of samples in both conditions, this approach will include one
#' additional sample in the comparison, potentially leading to improvement in
#' detection. However, the number of possible combinations with this approach
#' will be less than the ones possible with ERSSA. As a result, the identified
#' common gene list may not be as robust as the one identified with ERSSA.
#' Regardless, in practice, both approaches are likely to generate comparable
#' DE gene lists.
#'
#' @param deg The list of DE genes generated by one of ERSSA::DE_*.R scripts.
#' @param path Path to which the plot will be saved. Default to current working
#' directory.
#'
#' @return A list is returned containing:
#'  \itemize{
#'   \item{gg_object} {the ggplot2 object, which can then be further
#'   customized.}
#'   \item{intersect.dataframe} {the tidy table version used for plotting.}
#'   \item{intersect_genes} {list of vectors containing DE genes with vector
#'   name indicating the associated replicate level.}
#'   \item{full_num_DEG} {The number of DE genes with all samples included.}
#' }
#' @author Zixuan Shao, \email{Zixuanshao.zach@@gmail.com}
#'
#' @examples
#' # load edgeR deg object generated by erssa_edger using example dataset
#' # example dataset containing 1000 genes, 4 replicates and 10 comb. per rep.
#' # level
#' data(deg.partial, package = "ERSSA")
#'
#' gg_intersect = ggplot2_intersectPlot(deg.partial)
#'
#' @references
#' H. Wickham. ggplot2: Elegant Graphics for Data Analysis.
#' Springer-Verlag New York, 2009.
#'
#' @export


ggplot2_intersectPlot = function(deg=NULL, path='.'){

  if (is.null(deg)){
    stop('Missing required deg argument in ggplot2_intersectPlot function')
  }

  inters_genes = list() #to be populated with vectors of intersecting DE genes
  all_rep_levels = names(deg)[1:length(deg)-1] #all rep levels except full

  #calculate list of intersect genes
  for (rep_i in all_rep_levels){
    inters_genes_i = Reduce(intersect, deg[[rep_i]])
    inters_genes[[rep_i]] = inters_genes_i
  }

  full_num_DEG = length(deg$full$comb_1)

  #dataframe for plotting
  rep_levels = sapply(all_rep_levels, function(x)
      strsplit(x, split='_')[[1]][2])
  inters_genes_num = sapply(inters_genes, function(x) length(x))

  inters_genes_df = data.frame(replicate_number=rep_levels,
                           num_intersect_genes=inters_genes_num)
  inters_genes_df$replicate_number = factor(inters_genes_df$replicate_number,
                                            levels = unique(
                                              inters_genes_df$replicate_number))

  #plot
  gg = ggplot2::ggplot(inters_genes_df, ggplot2::aes_(x=~replicate_number,
                                                      y=~num_intersect_genes))+
    ggplot2::geom_col(width=0.7) +
    ggplot2::theme_bw(base_size=14) +
    ggplot2::geom_hline(ggplot2::aes(yintercept=full_num_DEG, color =
                                       "Full dataset DE genes"), size=0.75,
                        linetype="dashed", show.legend = TRUE) +
    ggplot2::scale_colour_manual(values=c("Full dataset DE genes"="blue")) +
    ggplot2::labs(x='Replicate number',
                  y='Number of intersecting DE genes', colour="") +
    ggplot2::geom_text(ggplot2::aes_(label = ~num_intersect_genes),
                       vjust = ifelse(inters_genes_df$num_intersect_genes >= 0,
                                      -0.2, 1.2))

  #create dir to save results
  folder_path = file.path(path)
  dir.create(folder_path, showWarnings = FALSE)


  #save plot and return ggplot2 object for additional customarization if needed
  ggplot2::ggsave(filename=file.path(path,'ERSSA_plot_3_IntersectDEGenes.png'),
         plot=gg, dpi=300, width = 20,
         height = 15, units = "cm")

  return(list(gg_object=gg, intersect.dataframe= inters_genes_df,
              intersect_genes= inters_genes, full_num_DEG=full_num_DEG))
}


